pkg/kubectl/resource_filter.go

```diff
@@ -0,0 +1,107 @@
+package kubectl
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// FilterFunc is a function that knows how to filter a specific resource kind.
+// It receives a generic runtime.Object which must be type-checked by the function.
+// Returns a boolean value true if a resource is filtered, or false otherwise.
+type FilterFunc func(runtime.Object, PrintOptions) bool
+
+// ResourceFilter filters api resources
+type ResourceFilter interface {
+	Filter(obj runtime.Object) (bool, error)
+	AddFilter(reflect.Type, FilterFunc) error
+	PrintFilterCount(io.Writer, string) error
+}
+
+// filterOptions is an implementation of ResourceFilter which contains
+// a map of FilterFuncs for given resource types, PrintOptions for a
+// resource, and a hiddenObjNum to keep track of the number of filtered resources
+type filterOptions struct {
+	filterMap    map[reflect.Type]FilterFunc
+	options      PrintOptions
+	hiddenObjNum int
+}
+
+func NewResourceFilter(opts *PrintOptions) ResourceFilter {
+	filterOpts := &filterOptions{
+		filterMap: make(map[reflect.Type]FilterFunc),
+		options:   *opts,
+	}
+
+	filterOpts.addDefaultHandlers()
+	return filterOpts
+}
+
+func (f *filterOptions) addDefaultHandlers() {
+	f.AddFilter(reflect.TypeOf(&api.Pod{}), filterPods)
+	f.AddFilter(reflect.TypeOf(&v1.Pod{}), filterPods)
+}
+
+func (f *filterOptions) AddFilter(objType reflect.Type, handlerFn FilterFunc) error {
+	f.filterMap[objType] = handlerFn
+	return nil
+}
+
+// filterPods is a FilterFunc type implementation.
+// returns true if a pod should be skipped. Defaults to true for terminated pods
+func filterPods(obj runtime.Object, options PrintOptions) bool {
+	var reason string
+
+	switch p := obj.(type) {
+	case *v1.Pod:
+		reason = string(p.Status.Phase)
+		if p.Status.Reason != "" {
+			reason = p.Status.Reason
+		}
+	case *api.Pod:
+		reason = string(p.Status.Phase)
+		if p.Status.Reason != "" {
+			reason = p.Status.Reason
+		}
+	}
+
+	if len(reason) == 0 {
+		return false
+	}
+
+	return !options.ShowAll && (reason == string(api.PodSucceeded) || reason == string(api.PodFailed))
```

Comment:
You're potentially comparing against `api` and `v1` here, while that works for now, it's not guaranteed to work in the future, if we ever change `api.PodSucceeded` (e.g. in `v2`)